### ISSUE TITLE
Update Read The Docs theme with upstream

### ISSUE THREAD

**Issue by d0ugal** [mkdocs/mkdocs#588]
The Upstream RTD theme has added some nice features for collapsing the navigation. You can see an example in [pip Sphinx documentation](http://pip.readthedocs.org). We have some users that could benifit from this, like [GlusterFS](http://gluster.readthedocs.org/en/latest/), as they have _many_ pages.

As a side, I happen to think that it is a negative update for smaller projects so I could be tempted to make this optional. Perhaps with an extra: config value like "extra: collapsable_nav: True" which would be specific to this theme? It could either remove the feature totally or expand by default.

**Comment by ProfYaffle** [comment]
Does this make more or less sense if combined with the 'first page heading appears in the ToC unless you suppress it' issue? Not sure they're related, but they are in my mind as I had to do the latter to try to address the length of the ToC, which is what this issue also addresses.

But - yes, please, optional or not!

**Comment by d0ugal** [comment]
It might make sense, yes. I've been pondering how to do that in a nice way.

**Comment by waylan** [comment]
This sounds like it would go in the template. What if it was in a block and/or include so it could be overridden by custom themes, etc. That strictly wouldn't require a setting, but given it makes more sense to have it off by default I can see you you might go that route...

What if it was extra css and/or extra JavaScript files that users could copy into their project if they want it--no setting required. At some point, I expect that various third parties will start to create and make available such things to add additional features. Might be nice to start a list of what's available and make this the first item on the list. That way, it is not included by default (no unnecessary bandwidth is used) but can be easily added.

**Comment by tterrag1098** [comment]
:+1: for this, would be very nice for the Forge docs project :)

**Comment by Collizo4sky** [comment]
Sorry fo hijacking this thread, what version of MkDocs is Readthedoc running? i can see build failing 'cos of older folder structure.

**Comment by ergonlogic** [comment]
:+1: for the Aegir Project docs too

**Comment by damonregan** [comment]
I too would like this collapsible navigation behavior seen at http://pip.readthedocs.org/en/stable/user_guide/ for my project documentation: http://sandboxdocs.readthedocs.org where the TOC is frustratingly long and loses focus when looking at pages at the bottom of the TOC (e.g., System APIs).  If the user clicks a System API page (e.g., Overview), the TOC goes back to the top of the page and the user can no longer see the System APIs section they were interested in reading.

**Comment by thenitai** [comment]
:+1:

**Comment by dmehra** [comment]
Another vote for collapsible nav option!

**Comment by Nightwolf42** [comment]
I would also like to have an option for a collapsible navigation menu. Showing all items at once looks totally chaotic with a large number of pages.

**Comment by thenitai** [comment]
Yes, totally confusing. We moved all our documentation over but showing all menus at once is chaotic. While we are at it, worse is that when you click on a menu it jumps all the way up, thus the user looses focus.

**Comment by brianjking** [comment]
@d0ugal @thenitai Any idea if there's a way to manually implement such a feature of the sphinx RTD theme with Mkdocs as it stands today? Thanks!

**Comment by thenitai** [comment]
@brianjking I'm sure there is a way to do it within the template somehow. I just haven't found the time to look into it, but this is definitely a must have. The RTD themes look nice on their side for other projects (not made with mkdocs) so I'm sure it is just a template issue. I can look into it, but would need a guide or alike...

**Comment by marcelstoer** [comment]
Just spent quite some time comparing my MkDocs setup with other projects (which are obviously Sphinx-based) because I also want those collapsible navigation items. Pity it's not supported.

**Comment by thenitai** [comment]
Hey guys,

I've taken the readthedocs template and spend some hours on this yesterday. I reverted to hidden div's and some jquery to make the menu expand and collapse. Additionally, I stored the state in a cookie as the template refreshed the whole page on each menu.

I know that the code could be made more streamlined but so far this all works quite good. You can see it in action over here at http://helpmonks.github.io/docs/

Let me know what you think.

**Comment by brianjking** [comment]
@thenitai Ooh, nice. Thanks.

* If I wanted to test it out it looks like I'd just include `theme_dir: readthedocshm` in my `mkdocs.yml` file and grab the folder `readthedocshm` from https://github.com/Helpmonks/docs/tree/master/readthedocshm and place it in my `mkdocs` folder, correct?

* I think I slightly prefer the icon to be on the left as seen below.

<img width="302" alt="screenshot2-19-1612 04" src="https://cloud.githubusercontent.com/assets/6786817/13184361/fa0c4cf2-d700-11e5-8bdf-4ad263e1dc37.png">

* What are your thoughts on having the arrow pointing up/down/etc based on the collapsed/expanded state of that element?

**Comment by thenitai** [comment]
@brianjking Yes, simply grab the one in the master. It is configured as custom themes should be configured according to the mkdocs. You can see it in the mkdocs.yml file.

I was thinking of the icon on the left and making the icon reflect on the state. This can be easily be achieved by some more jquery magic :)

You are invited to fork it and make a pull request. Though, I would love to make the tweak depend on variables. I believe we could utilize the "extra:" in the config for that, so we can have some options.

**Comment by d0ugal** [comment]
@thenitai Neat! One slight bug/issue. If I link somebody directly to a page, for example [this one](http://helpmonks.github.io/docs/api/customer_find/) the navigation will be collapsed despite them being on a sub page.

**Comment by thenitai** [comment]
@d0ugal Yes, I know. I haven't figured out how to capture it when accessing a link directly. It would be handy, though. I'm not Python/Jinja2 guy (actually my first stab at it). I'm at home with node.js/jade :) So any help appreciated.

**Comment by d0ugal** [comment]
@thenitai The *active* menu item is a `<li>` with the class `current`. So, with that you should be able to keep going through it's parents and making sure they are visible until you each the top of the nav?

**Comment by thenitai** [comment]
@d0ugal Ah true (it was very late last night)... I'm on my way to a week vacation but will look into it. It's something that I definitely want.... and sorry for hijacking this thread.

**Comment by d0ugal** [comment]
No worries, it could be that your work is a starting point towards this larger update.

**Comment by thenitai** [comment]
Ok, just let me know please so I'm not doing the double work here :)

**Comment by claygorman** [comment]
+1

**Comment by hartzell** [comment]
I grabbed a copy of the theme as described by @brianjking and it's working, but:

- there's a very noticeable flash in the Table of Contents every time I move to a new section; and
- I don't see any expandable sections, its not clear if I just need to test it on a bigger document or ...

Thanks for doing the work.

**Comment by hartzell** [comment]
I got the little arrows working, I wasn't going deep enough with section headers (I had screwed up and tried to go deep with items that had files associated with them).

**Comment by tterrag1098** [comment]
Any progress on this? Still very interested in this change.

**Comment by d0ugal** [comment]
@tterrag1098 This issue is waiting for somebody with enough time and motivation to come along and put the work in.

**Comment by tterrag1098** [comment]
It seems like that has already started. I guess I was asking those people :P

**Comment by d0ugal** [comment]
@tterrag1098 I see, in that case I don't know if they have progressed with this outside of MkDocs.

**Comment by moth3r** [comment]
+1 for that functionality be implemented!

**Comment by wmarshall484** [comment]
+1

I would like this functionality for a project I'm working on. Unfortunately, it seems I can't use mkdocs without the collapsible sidebar. Are there any updates?

**Comment by thenitai** [comment]
@wmarshall484 You can use the template that I provided. Scroll up to see my comment and the link to the template. It's rather easy to set up.

**Comment by wmarshall484** [comment]
@thenitai, thank you! It was very easy to set up. 

Comparing the looks of the most recent Sphinx version (found [here](http://docs.readthedocs.io/en/latest/builds.html#how-we-build-documentation))
![p2](https://cloud.githubusercontent.com/assets/4752570/17346672/be77f974-58c0-11e6-912c-d1e9c8ce0b09.JPG)

 to @thenitai's version:
![p1](https://cloud.githubusercontent.com/assets/4752570/17346679/c49e93da-58c0-11e6-91c9-2d0802275083.JPG)

I like Sphinx's use of grey shading to denote depth, as well as the [-] icons which denote expanded tables. The latter version highlights the bars as white at every expanded layer, which I don't think is as readable. The fix should work for now, but it would be great if the Sphinx functionality could make it into master. Were I a web developer who knew up from down, I would absolutely contribute this :)

**Comment by thenitai** [comment]
@wmarshall484 hhmhmmh.... I can definitely look into it. However, I'm quite busy these days, so I'm not sure when I get to it.

**Comment by MaciejCaputa** [comment]
I will have a look at this as well since I would also prefer having a functionality for expandable menu that is similar to *Sphinx's*.

**Comment by JosephSun** [comment]
@thenitai   
Thanks for doing this. What the max nested folder structure you can have? I currently can only get to 2 nested folder structures. Once I click on the the 3rd nested directory, the entire side nav bar collapses. 

![mkdocs nested folder](https://cloud.githubusercontent.com/assets/11227083/18724066/098eaf1a-8009-11e6-946a-e185bbe18c88.PNG)

**Comment by naushadjamil** [comment]
@thenitai and anybody else in this thread looking for a bug free functional collapsible navigation theme. I need something like the sphinx RTD theme with the +/- thing (more importantly that theme seems to work without issues). The one @thenitai crafted seems to have several issues. I am stuck with my documentation site at the point where I need to again invest a good deal of time in another static site generator that supports collapsible navigation (e.g. sphinx itself). Does anybody have any idea what can be done here?

**Comment by waylan** [comment]
Just to make sure there is no confusion here; the maintainers of MkDocs do not personally use the `readthedocs` theme. Therefore, we are more inclined to dedicate our time and resources on other aspects of MkDocs. That said, we are happy to keep the theme updated if others are willing to do the work. Therefore, for those who care about it, the best way to move things forward is to create pull requests referencing this issue. We are happy to review such code submissions by others.

If you are concerned that your work may not be accepted, that is understandable. A good rule of thumb with the `readthedocs` theme is that if your contribution is adding/updating the theme to better match the Sphinx theme of the same name, then we would be willing to accept it in most cases (pending a review for quality/breakage/etc). Adding things not included in the Sphinx theme of the same name are generally not going to be accepted without good reason and should probably be discussed first.

I mention this because there has been a lot of discussion since this issue was opened but to-date no one has actually submitted any pull requests. I'm not expecting one big code dump which fixes everything at once. Small piecemeal pull requests implementing one feature at a time would actually be preferred.

**Comment by thenitai** [comment]
@naushadjamil you mentioned that my template "seems to have several issues". Can you be more specific and what is not working for you? As general "doesn't work" is not helpful at all.

For reference the theme is active and working at http://developer.helpmonks.com/

**Comment by naushadjamil** [comment]
@thenitai - my doc tree is pretty deep (easily 4 or more). somebody already reported above one issue with deep nesting. First when I expand an innermost item I don't see the leaf page there. Second if there are two or more root level trees expanding the 2nd root expands all trees. There are a couple other glitches as well. I recommend you test with a document tree with deeper nesting. Let me know if you want to see snapshots or a live site. The other issue is visual only not really an issue. If not +/- at least something that shows the expanded/collapsed status of a tree. Is it simple to tweak it to show +/- and on the left side?

**Comment by naushadjamil** [comment]
@waylan - It isn't about the theme so much as it is about functionality. I will claim MkDocs doesn't support my documentation scenario. My document tree is deep (say 5) and there are 100s of root level folders. Rendering them with menu based navigation doesn't work for two reasons. One - difficult to navigate using deep menu that constantly pops up more menus. Two - eventually it rolls off the screen and user cannot see the document. ReadTheDocs theme was the only one that actually was able to render it only problem is with all 100s of folder with tons of nested folders when they are expanded it is pretty useless.

Hence the claim. We can't support a case for say 100 folders of documentation with each as deep as 5 levels.

I don't think I gave an idea that I will be willing to do any dev work on any of this. I am just an end user trying to create a documentation site with MkDocs. Or do you expect most of your users to actually contribute to MkDocs sources in some way?

There were tons of static site generators. But my Google search showed me MkDocs at the top and that's what I picked and spent a good deal of time with and now got stuck with my scenario.

Thank you.

**Comment by naushadjamil** [comment]
@thenitai You should probably make a pull request for whatever you have. I think the work should be done in a separate theme instead of updating readthedocs theme because as is it works now. Just a new theme based on readthedocs that supports collapsible doctree.

**Comment by waylan** [comment]
@naushadjamil my comment was directed at those who have already done some work (or are inclined to). I was attempting to encourage them to contribute the work back (if they desire, no pressure) so we can move this issue forward. It's something I've been meaning to do, your comment just reminded me about it.

By the way, there seems to be a small subset of users who report problems connected with a very large number of documents. We don't really test for that and as far as I can tell, most of our users don't need support for that. It may be that a different static site generator works better for large collections of documents. But I wouldn't know as I've never had a need for that use case.

**Comment by naushadjamil** [comment]
@waylan thanks for clarification.

**Comment by marcelstoer** [comment]
> I think the work should be done in a separate theme instead of updating readthedocs theme because as is it works now.

A theme called 'readthedocs' should look and work just like the Read the Docs standard theme - including collapsible navigation. So, I'd hate to see a new theme that looks similar, has collapsible navigation but different than Read the Docs.

**Comment by d0ugal** [comment]
The themes goal is indeed to be as similar to the Sphinx theme as is reasonable. Sphinx and MkDocs do not match up in terms of features, so some departure is expected but they should generally feel the same.

**Comment by naushadjamil** [comment]
@theintai - if you could point me to what needs to be done I can take a stab at it.

**Comment by thenitai** [comment]
@naushadjamil Thanks. I simply don't have time these days. Easiest is to fork the theme and issue pull requests. If you look at the code, I think it's rather easy to figure it out. It's just a bunch of Javascript/jquery that listens to some events. Let me know if you get stuck.

**Comment by waylan** [comment]
FYI, I just pushed a refactor of the navigation in the `readthedocs` theme to #1042. I suspect the work to get the navigation updated to upstream would be a little simpler once that PR is merged. I haven't checked, but it may be as simple as updating the CSS and JavaScript to that used in [snide/sphinx_rtd_theme](https://github.com/snide/sphinx_rtd_theme).

**Comment by naushadjamil** [comment]
@waylan thanks for letting us know there is a bit of work towards getting the collapsible navigation in the readthedocs theme. Looks like you are still unsure when or if it will ever match the sphinx_rtd_theme anytime soon. If this happens please let us know. Meanwhile, I need to explore other static site generators to see if I can get collapsible tree like navigation.

**Comment by laoshaw** [comment]
same here, need the collapsible navigation as my site has a long sidebar for the navigation.

**Comment by naushadjamil** [comment]
@waylan any progress on that collapsible navigation tree support. I am looking into Hugo. It seems they don't have built-in search.

**Comment by waylan** [comment]
@naushadjamil trying to shoehorn a feature in when the upstream project doesn't support it has met with a few roadblocks as outlined in #1042. You can read the details there. In the end, I expect that that won't ever see the light of day, but much of the work I did on the readthedocs theme will get ripped out and used as a better navigation template. The existing template doesn't even produce valid HTML.

**Comment by hartzell** [comment]
Folks who are looking for collapsing sidebars might find what they want in @tomjohnson1492's documentation theme for Jekyll, here: https://github.com/tomjohnson1492/documentation-theme-jekyll

If you want to use different icons for the sidebar, there's an example in [this PR](https://github.com/tomjohnson1492/documentation-theme-jekyll/pull/65).

It's not mkdocs.  It is Jekyll and has the advantage of being GitHub Pages compatible.

Whether it works for you is for you to figure out.

**Comment by evitalis** [comment]
Has there been any further development on this issue? I am looking forward to using the theme but once I get enough documents this will be a mess since the headers are all expanded.

**Comment by waylan** [comment]
I can't speak for @d0ugal but I'm no CSS master and am hoping someone volunteers their time to work on this. I don't use this theme myself and have other needs/priorities. As a reminder, the maintainers are volunteers working on this in their spare time so we tend to focus on what is important to us. If you have a pet feature/bug the best way to get it addressed is to do the work yourself. I'm always happy to review pull requests, especially if they resolve bugs and/or longstanding and popular feature requests.

**Comment by d0ugal** [comment]
I am also lacking in front end skills. I think this is one of the biggest issues with MkDocs and the current regular contributors. The best CSS work I see happening in the MkDocs community is in the Material theme.

So far my attempts to find somebody that is interested, able and willing to help have failed.

**Comment by waylan** [comment]
For those who are following this feature, I created #1155 which is a first step by making improvements to the underlying template. I think that the rest of this can be completed with CSS/JS work only.

However, in the comments for #1155 I also outline the differences that still exist in the HTML between our theme and the upstream (Sphinx) theme. If those changes are necessary (to complete the request), they would introduce backward incompatible changes to MkDocs, in which case I suggest the theme be forked under a new name. Any feedback is welcome.

**Comment by aldanor** [comment]
Collapsable headings and other niceties aside, the real show-stopper with the RTD theme is the lack support for nested headings in TOC. So if you have a document with a lot of headings from H1-H6, they will all get dumped inline into a big mess of section headings. Is there an easy way to support that?

**Comment by cjsheets** [comment]
For anyone interested, I setup a version of the mkdocs ReadTheDocs theme with collapsible sidebar support. You can see the theme [in action here](http://readthedocs.sheets.ch), or see [installation instructions](https://github.com/cjsheets/mkdocs-rtd-dropdown).

![Example](https://raw.githubusercontent.com/cjsheets/mkdocs-rtd-dropdown/master/docs/img/example.png)

The theme is stable though still under development. Still needs a number of refinements and I don't know if it supports all MkDocs file structures.

As waylan mentioned (here and in other issues), there are differences between how mkdocs and readthedocs parse files/headings so a direct clone of ReadTheDocs behavior would require changes to the mkdocs core. I didn't want to do that because I don't want to maintain a fork of the entire project.

For my purposes, a RTD-style theme that collapses the sidebar to a manageable size is sufficient.

**Comment by bastianeicher** [comment]
I put together a few lines of JavaScript that collapse menu sub-sections in MkDocs's ReadTheDocs theme. It also auto-scroll the menu to the active page.

```javascript
function collapse(tree) {
  $(tree).children("span.caption-text").addClass("expandable");
  $(tree).children("ul.subnav").hide();
}

function expand(tree) {
  $(tree).children("span.caption-text").removeClass("expandable");
  $(tree).children("ul.subnav").show();
}

// Collapse all sub-menus except the containing the current page
$("li.toctree-l1").each(function() {
  if ($(this).find(".current").length == 0) {
    collapse(this);
  }
});

// Scroll to the current page
$("nav").scrollTop($("a.current").offset().top - 128);

// Expand sub-menu on click and collapse all other sub-menus
$("span.caption-text").click(function(event) {
  if ($(event.target).hasClass("expandable")) {
    collapse("li.toctree-l1");
    expand($(event.target).parent());
  }
});
```
(https://github.com/0install/docs/blob/master/docs/menu.js)

It requires the following additional CSS:
```css
span.expandable { cursor:pointer; }
```

You can put these into files named `menu.js` and `style.cs` in your `docs` directory and reference them from `mkdocs.yml`:
```yml
extra_css: [style.css]
extra_javascript: [menu.js]
```

You can see the result at: https://0install.github.io/docs/

**Comment by cjsheets** [comment]
Awesome! I'll check it out.

If it looks like "rtd dropdown"s behavior is covered by the native collapsible nav support, I'll probably just depreciate the theme and advise to use the new feature 😃

**Comment by NateZimmer** [comment]
So its 5/2019 and this still isn't out. How does one use mkdocs 1.1? Will mkdocs 1.1 ever be released? Current release is 1.04

**Comment by xgaia** [comment]
Hello everyone!  Is there a release date for the 1.1 release? I'm interested in the dropdown menu functionality.

**Comment by waylan** [comment]
I work on MkDocs as a volunteer in my spare time, which is very short supply. I don't have time to do all of the work required to properly prepare a release. If anyone wants to help you have two options:

1. Provide fixes (PRs) for the issues listed in [Milestone 1.1][1]. When those all get fixed (especially #1865), then we can release 1.1. If this gets done, I'll find the time to review the PRs and, when they are completed, do the release.
2. Backport all of the bug fixes listed in the [Release Notes][2] for version 1.1 (specifically only the items which start with "Bugfix: ...") to the `1.0` branch. All of the relevant commits can be included in one PR for this. If this gets done, I'll find the time to do a 1.0.5 release with the bug fixes (including the fix for this issue).

However, until either of those two things happen, there will be no new releases. If I find some unexpected free time in the future, I'll do the work, but I don't expect to anytime soon. Any help you can provide would be appreciated.

[1]: https://github.com/mkdocs/mkdocs/milestone/1.1
[2]: https://github.com/mkdocs/mkdocs/blob/262c2b70f3b1a450d685530610af3f28e12f9c9f/docs/about/release-notes.md#other-changes-and-additions-to-version-11
